kernel/src/syscall.rs

```diff
@@ -69,35 +154,239 @@ impl Syscall {
+/// [`encode_syscall_return`](GenericSyscallReturnValue::encode_syscall_return)
+/// method.
+///
+/// Capsules use higher level Rust types
+/// (e.g. [`AppSlice`](crate::AppSlice) and
+/// [`Callback`](crate::Callback)) or wrappers around this struct
+/// ([`CommandResult`](crate::CommandResult)) which limit the
+/// available constructors to safely constructable variants.
+#[derive(Copy, Clone, Debug)]
+pub enum GenericSyscallReturnValue {
+    /// Generic error case
+    Failure(ErrorCode),
+    /// Generic error case, with an additional 32-bit data field
+    FailureU32(ErrorCode, u32),
+    /// Generic error case, with two additional 32-bit data fields
+    FailureU32U32(ErrorCode, u32, u32),
+    /// Generic error case, with an additional 64-bit data field
+    FailureU64(ErrorCode, u64),
+    /// Generic success case
+    Success,
+    /// Generic success case, with an additional 32-bit data field
+    SuccessU32(u32),
+    /// Generic success case, with two additional 32-bit data fields
+    SuccessU32U32(u32, u32),
+    /// Generic success case, with three additional 32-bit data fields
+    SuccessU32U32U32(u32, u32, u32),
+    /// Generic success case, with an additional 64-bit data field
+    SuccessU64(u64),
+    /// Generic success case, with an additional 32-bit and 64-bit
+    /// data field
+    SuccessU64U32(u64, u32),
+
+    // These following types are used by the scheduler so that it can
+    // return values to userspace in an architecture (pointer-width)
+    // independent way. The kernel passes these types (rather than
+    // AppSlice or Callback) for two reasons. First, since the
+    // kernel/scheduler makes promises about the lifetime and safety
+    // of these types (e.g., an accepted allow does not overlap with
+    // an existing accepted AppSlice), it does not want to leak them
+    // to other code. Second, if subscribe or allow calls pass invalid
+    // values (pointers out of valid memory), the kernel cannot
+    // construct an AppSlice or Callback type but needs to be able to
+    // return a failure. -pal 11/24/20
+    /// Read/Write allow success case
+    AllowReadWriteSuccess(*mut u8, usize),
+    /// Read/Write allow failure case
+    AllowReadWriteFailure(ErrorCode, *mut u8, usize),
+
+    /// Read only allow success case
+    AllowReadOnlySuccess(*const u8, usize),
+    /// Read only allow failure case
+    AllowReadOnlyFailure(ErrorCode, *const u8, usize),
+
+    /// Subscribe success case
+    SubscribeSuccess(*const u8, usize),
+    /// Subscribe failure case
+    SubscribeFailure(ErrorCode, *const u8, usize),
+}
+
+impl GenericSyscallReturnValue {
+    pub(crate) fn from_command_result(res: CommandResult) -> Self {
+        res.into_inner()
+    }
+
+    /// Encode the system call return value into 4 registers
+    ///
+    /// Architectures are free to define their own encoding.
+    pub fn encode_syscall_return(&self, a0: &mut u32, a1: &mut u32, a2: &mut u32, a3: &mut u32) {
```

Comment:
A little more description would be helpful here. Specifically, why not just have this function return `(r0: u32, r1: u32, r2: u32, r3: u32)` and let the caller decide what to do with those values?

Reply:
Returning a tuple was actually my original proposal way back in some email. :) Making `encode_syscall_return` take mutable references to the memory being modified was an idea of Phil, which I've quite liked. It guarantees us that we're, in the encoding function, directly writing to the target location without risk of copying the values on the stack temporarily and not having to think about compiler optimization.

However, this makes dealing with types of potentially(!) different size really complex (since one can't simply transmute a `&mut usize` into a `&mut u32`). I'll see whether this has any real performance impact and change it back to returning a tuple if not.

Reply:
It means we don't have to copy the values onto the stack, then copy them elsewhere.